pkg/generator/gardenlinux/cgroups.go

```diff
@@ -0,0 +1,168 @@
+	"github.com/gardener/gardener/pkg/utils/version"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	// defaultCgroup is the cgroup version to fall back to
+	defaultCgroup = v1alpha1.CgroupVersion(v1alpha1.CgroupVersionV1)
+
+	// cgroupUnitContent is the content of the systemd unit to start the cgroup configuration script
+	cgroupUnitContent string = `[Unit]
+Description=Configure cgroup version for Gardener
+After=cloud-config-downloader.service
+Before=gardener-restart-system.service kubelet.service
+
+[Install]
+WantedBy=multi-user.target
+
+[Service]
+Type=oneshot
+ExecStart=/opt/gardener/bin/configure_cgroups.sh
+RemainAfterExit=true
+StandardOutput=journal
+`
+
+	// containerdDropInContent is the content of the systemd drop-in to configure the cgroup driver for containerd
+	containerdDropInContent string = `[Service]
+ExecStartPre=/opt/gardener/bin/containerd_cgroup_driver.sh
+`
+
+	// kubeletDropInContent is the content of the systemd drop-in to configure the cgroup driver for kubelet
+	kubeletDropInContent string = `[Service]
+ExecStartPre=/opt/gardener/bin/kubelet_cgroup_driver.sh
+`
+)
+
+func ConfigureCgroups(ctx context.Context, client client.Client, osc *extensionsv1alpha1.OperatingSystemConfig, decoder runtime.Decoder) (*generator.File, *generator.Unit, error) {
+	cgroupVersion, err := extractCgroupVersion(ctx, client, osc, decoder)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	config := map[string]interface{}{
+		"cgroupVersion": cgroupVersion,
+	}
+
+	var buff bytes.Buffer
+	cgroupScriptTemplate, err := templates.ReadFile(filepath.Join("templates", "configure_cgroups.sh.tpl"))
+	if err != nil {
+		return nil, nil, err
+	}
+	t, err := template.New("cgroupScript").Parse(string(cgroupScriptTemplate))
+	if err != nil {
+		return nil, nil, err
+	}
+	if err = t.Execute(&buff, config); err != nil {
+		return nil, nil, err
+	}
+
+	cgroupScript := &generator.File{
+		Path:        filepath.Join(scriptLocation, "configure_cgroups.sh"),
+		Content:     buff.Bytes(),
+		Permissions: &scriptPermissions,
+	}
+
+	cgroupUnit := &generator.Unit{
+		Name:    "gardener-configure-cgroups.service",
+		Content: []byte(cgroupUnitContent),
+	}
+
+	return cgroupScript, cgroupUnit, nil
+}
+
+func extractCgroupVersion(ctx context.Context, client client.Client, osc *extensionsv1alpha1.OperatingSystemConfig, decoder runtime.Decoder) (*v1alpha1.CgroupVersion, error) {
+	providerConfig := osc.Spec.ProviderConfig
+
+	if providerConfig != nil {
+		obj := &v1alpha1.OperatingSystemConfiguration{}
+
+		if _, _, err := decoder.Decode(providerConfig.Raw, nil, obj); err != nil {
+			return nil, fmt.Errorf("failed to decode provider config: %+v", err)
+		}
+
+		shoot, err := extensionscontroller.GetShoot(ctx, client, osc.Namespace)
+		if err != nil {
+			return nil, err
+		}
+
+		shootKubernetesAtLeast119, err := version.CompareVersions(shoot.Spec.Kubernetes.Version, ">=", "1.19")
```

Comment:
@gardener/gardener-maintainers what do you think about deprecating and removing support for k8s <=1.18 for shoot cluster?

Reply:
> I think you will need to check also the k8s version of the worker pool as it could differ from the control plane k8s version, ref: https://github.com/gardener/gardener/blob/master/docs/usage/worker_pool_k8s_versions.md
> 
> The worker pool name is set in a label in the OSC resource (ref: [gardener/gardener#5256](https://github.com/gardener/gardener/pull/5256)), so you should be able to easily find the worker config for the given operatingSystemConfig.

That has been taken care of.